ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3MultipartCompleteUploadHandler.java

```
@@ -415,4 +430,42 @@ List<ChunkInfo> getChunksToStitch(CompleteMultipartUpload completeMultipartUploa
       return chunkInfos;
     }
   }
+
+  /**
+   * Check the list size and part number before processing request
+   * 1. Disallow duplicate part numbers
+   * 2. Disallow duplicate etags
+   * 3. Check for list size 10000
+   * 4. Check for part numbers integer 1-10000
+   * @param parts sorted parts list
+   * @return the bad request error
+   */
+  private static String validateParts(List<Part> parts) {
+    if (parts.size() > MAX_LIST_SIZE) {
+      return String.format("Parts list size cannot exceed {}.", MAX_LIST_SIZE);
```

Throw an exception instead of return a string.
```
          throw new RestServiceException(message, RestServiceErrorCode.BadRequest);

```

---

I think the current logic: return string and if the string is not null, log & throw error is more concise.

---

Take a look at some of the tickets we get. Log messages are usually redacted or not printed. The only thing to go fwd are teh stack traces. So that must be accurate to the line of code.

---

okay